src/SMP.cpp

```diff
@@ -26,19 +27,31 @@ SMP::Mutex::Mutex() {
 
 SMP::Lock::Lock(Mutex & m) {
     m_mutex = &m;
+    m_selflock = false;
     lock();
 }
 
 void SMP::Lock::lock() {
+    assert(!m_selflock);
     while (m_mutex->m_lock.exchange(true, std::memory_order_acquire) == true);
+    m_selflock = true;
 }
 
 void SMP::Lock::unlock() {
-    m_mutex->m_lock.store(false, std::memory_order_release);
+    assert(m_selflock);
+    auto prevval = m_mutex->m_lock.exchange(false, std::memory_order_release);
+
+    // if this fails it means we are unlocking an unlocked lock
+    assert(prevval);
+    m_selflock = false;
```

Comment:
How can we be sure that `m_selflock` has the right value if one thread just unlocked the mutex and another thread locked the mutex before this line was reached?

Reply:
m_selflock is only modified by the mutex itself, so it can never be modified concurrently.